### DISCUSSION TITLE
Broker multi-user problem

### DISCUSSION THREAD

**Discussion by zcz-1020** [submariner-io/submariner#1991] [Q&A]
Does the broker support multiple users， and each user has its own clusterset?

**Comment by skitt** [comment]
The broker currently only supports one clusterset.

Note that while it would be possible to provide multiple clustersets, the ramifications are complex (especially for gateways) and the MCS API is designed around clusters belonging to a single clusterset.

**Reply by zcz-1020** [reply]
Thank you for your answer. I still have a question that I have not figured out how to maintain and save the clusters in the clusterset（i.e., which clusters are in the clusterset）.  I did not find the corresponding code in the source code, nor did I see the clusterset object.

**Reply by tpantelis** [reply]
Submariner itself does not have the concept of a clusterset object as you described.

**Comment by mkolesnik** [comment]
Submariner doesn't deal with "users", it should be possible to deploy two (or more) different brokers on the same cluster (broker is mainly a namespace + CRDs + RBAC) and craft RBAC so that each user can only access their own broker.

As Stephen said it's complex, but possible.

**Reply by tpantelis** [reply]
There is also an operator pod that runs in the `submariner-operator` namespace if deployed via `subctl`. It looks like it **should** work with multiple brokers but I'm not sure that's ever been tested.

**Reply by zcz-1020** [reply]
Thank you for your answer. I also think that I can create multiple brokers to manage multiple clustersets. Each broker corresponds to a clusterset for isolation. It's really complicated，and it is not clear whether it will actually work.

**Reply by zcz-1020** [reply]
However, I think of a question. Multiple brokers share an ectd backend storage. In this case, will the CRDs watched by the cluster be affected?

**Reply by skitt** [reply]
Joined clusters watch a single namespace in the broker cluster; if each clusterset has its own namespace, the watched objects will be kept separate.

**Reply by zcz-1020** [reply]
> Joined clusters watch a single namespace in the broker cluster; if each clusterset has its own namespace, the watched objects will be kept separate.

Thank you very much. I understand. Are you advised to create multiple brokers and each broker has its own namespace or one broker has multiple namespaces so that each clusterset corresponds to a namespace?

**Reply by skitt** [reply]
One broker effectively _is_ a namespace; to create multiple brokers, you create multiple namespaces, and configure the joined clusters to use the appropriate broker namespace.

**Reply by zcz-1020** [reply]
I understand, thank you very much!
